plugins/corsair/fu-corsair-bp.c

```diff
@@ -0,0 +1,407 @@
+{
+	guint8 commit_cmd[FU_CORSAIR_MAX_CMD_SIZE] = {0x08, 0x05, 0x01, 0x00};
+	if (!fu_corsair_bp_command(self, commit_cmd, CORSAIR_TRANSACTION_TIMEOUT, TRUE, error)) {
+		g_prefix_error(error, "firmware commit fail: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+gboolean
+fu_corsair_bp_write_firmware(FuCorsairBp *self,
+			     FuFirmware *firmware,
+			     FuProgress *progress,
+			     GError **error)
+{
+	const guint8 *firmware_raw;
+	gsize firmware_size;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+	g_autoptr(FuChunk) firstChunk = NULL;
+	g_autoptr(GBytes) rest_of_firmware = NULL;
+	guint32 first_chunk_size = self->cmd_write_size - CORSAIR_FIRST_CHUNK_HEADER_SIZE;
+
+	blob = fu_firmware_get_bytes(firmware, error);
+	if (blob == NULL) {
+		g_prefix_error(error, "cannot get firmware data");
+		return FALSE;
+	}
+	firmware_raw = fu_bytes_get_data_safe(blob, &firmware_size, error);
+	if (firmware_raw == NULL) {
+		g_prefix_error(error, "cannot get firmware data: ");
+		return FALSE;
+	}
+
+	/* the firmware size should be greater than 1 chunk */
+	if (firmware_size <= first_chunk_size) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "update file should be bigger");
+		return FALSE;
+	}
+
+	firstChunk = fu_chunk_new(0, 0, 0, g_bytes_get_data(blob, NULL), first_chunk_size);
+	rest_of_firmware = fu_common_bytes_new_offset(blob,
+						      first_chunk_size,
+						      firmware_size - first_chunk_size,
+						      error);
+	if (rest_of_firmware == NULL) {
+		g_prefix_error(error, "cannot get firmware past first chunk: ");
+		return FALSE;
+	}
+	chunks =
+	    fu_chunk_array_new_from_bytes(rest_of_firmware,
+					  first_chunk_size,
+					  0,
+					  self->cmd_write_size - CORSAIR_NEXT_CHUNKS_HEADER_SIZE);
+
+	if (!fu_corsair_bp_write_firmware_chunks(self,
+						 firstChunk,
+						 chunks,
+						 progress,
+						 g_bytes_get_size(blob),
+						 error))
+		return FALSE;
+
+	if (!fu_corsair_bp_commit_firmware(self, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+gboolean
+fu_corsair_bp_activate_firmware(FuCorsairBp *self, FuFirmware *firmware, GError **error)
+{
+	guint32 crc;
+	gsize firmware_size;
+	const guint8 *firmware_raw;
+	g_autoptr(GBytes) blob = NULL;
+	guint8 cmd[FU_CORSAIR_MAX_CMD_SIZE] = {0x08, 0x16, 0x00, 0x01, 0x03, 0x00, 0x01, 0x01};
+
+	blob = fu_firmware_get_bytes(firmware, error);
+	if (blob == NULL) {
+		g_prefix_error(error, "cannot get firmware bytes");
+		return FALSE;
+	}
+
+	firmware_raw = fu_bytes_get_data_safe(blob, &firmware_size, error);
+	if (firmware_raw == NULL) {
+		g_prefix_error(error, "cannot get firmware data: ");
+		return FALSE;
+	}
+
+	crc = fu_corsair_calculate_crc(firmware_raw, firmware_size);
+	fu_common_write_uint32(&cmd[CORSAIR_OFFSET_CMD_CRC], crc, G_LITTLE_ENDIAN);
+
+	return fu_corsair_bp_command(self, cmd, CORSAIR_ACTIVATION_TIMEOUT, TRUE, error);
+}
+
+gboolean
+fu_corsair_bp_attach_legacy(FuCorsairBp *self, GError **error)
```

Comment:
if you derive from FuUsbDevice you can use the ->attach() vfunc

Reply:
If it helps, fu_device_attach() is a thin wrapper -- so you can unexport this, and then just do `fu_device_attach(bp_device)` if you connect up the vfunc